Validation/HLTrigger/src/HLTGenValHistCollFilter.cc

```diff
@@ -0,0 +1,143 @@
+  dR2limit_ = filterCollConfig.getParameter<double>("dR2limit");
+}
+
+edm::ParameterSetDescription HLTGenValHistCollFilter::makePSetDescription() {
+  edm::ParameterSetDescription desc;
+  desc.add<std::string>("objType", "");
+  desc.add<std::string>("hltProcessName", "HLT");
+  desc.add<double>("dR2limit", 0.1);
+  desc.add<std::string>("filterName", "");
+  return desc;
+}
+
+// general hist booking function, receiving configurations for 1D and 2D hists and calling the respective functions
+void HLTGenValHistCollFilter::bookHists(DQMStore::IBooker& iBooker, std::vector<edm::ParameterSet>& histConfigs, std::vector<edm::ParameterSet>& histConfigs2D) {
+  for (auto& histConfig : histConfigs) book1D(iBooker, histConfig);
+  for (auto& histConfig : histConfigs2D) book2D(iBooker, histConfig);
+}
+
+// histogram filling routine
+void HLTGenValHistCollFilter::fillHists(const HLTGenValObject& obj, edm::Handle<trigger::TriggerEvent>& triggerEvent) {
+
+  // this handles the "before" step, denoted by a "dummy" filter called "beforeAnyFilter"
+  // the histogram is filled without any additional requirements for all objects
+  if(filter_ == "beforeAnyFilter") {
+    for (auto& hist : hists_) hist->fill(obj);
+  } else {
+    // main filling code
+
+    // get filter object from filter name
+    edm::InputTag filterTag(filter_, "", hltprocessname_);
+    size_t filterIndex = triggerEvent->filterIndex(filterTag);
+
+    // get trigger objects passing filter in question
+    trigger::TriggerObjectCollection allTriggerObjects = triggerEvent->getObjects(); // all objects
+    trigger::TriggerObjectCollection selectedObjects; // vector to fill with objects passing our filter
+    if (filterIndex < triggerEvent->sizeFilters()) {
+      const auto& keys = triggerEvent->filterKeys(filterIndex);
+      for (unsigned short key : keys) {
+        trigger::TriggerObject foundObject = allTriggerObjects[key];
+        selectedObjects.push_back(foundObject);
+      }
+    }
+
+    // differentiate between event level and particle level variables
+    std::vector<std::string> event_level_variables = {"AK4HT", "AK8HT", "MET"};
+    if(std::find(event_level_variables.begin(), event_level_variables.end(), objType_) != event_level_variables.end()) {
+      // for these event level variables we only require the existence of a trigger object, but no matching
+      if(selectedObjects.size() > 0) for (auto& hist : hists_) hist->fill(obj);
+    } else {
+      // do a deltaR matching between trigger object and GEN object
+      double mindR2 = 99999;
+      for (const auto & filterobj : selectedObjects) {
+        double dR = deltaR2(obj, filterobj);
+        if(dR < mindR2) mindR2 = dR;
+      }
+      if(mindR2 < dR2limit_) for (auto& hist : hists_) hist->fill(obj);
+    }
+
+  }
+}
+
+// booker function for 1D hists
+void HLTGenValHistCollFilter::book1D(DQMStore::IBooker& iBooker, edm::ParameterSet& histConfig) {
```

Comment:
histConfig should be constant (unless you are modifying it somehow which would be werid...)